File: data structures and algorithms/java/05_002.java
import java.util.*;
import java.util.stream.*;
import java.io.*;
import java.math.*;

public class Main {
    public static void main(String[] args) {
        // Main method code goes here
    }
    
    public class BookRecommendationGraph {
        private Map<String, Map<String, Integer>> graph;
        
        public BookRecommendationGraph() {
            this.graph = new HashMap<>();
        }
        
        public void addBookRelationship(String book1, 
                                        String book2, 
                                        int weight) {
            if (!this.graph.containsKey(book1)) {
                this.graph.put(book1, new HashMap<>());
            }
            if (!this.graph.containsKey(book2)) {
                this.graph.put(book2, new HashMap<>());
            }
                
            this.graph.get(book1).put(book2, weight);
            this.graph.get(book2).put(book1, weight);
        }
        
        public List<Map.Entry<String, Integer>> 
        getRecommendations(String book, int limit) {
            if (!this.graph.containsKey(book)) {
                return new ArrayList<>();
            }
                
            List<Map.Entry<String, Integer>> 
                sortedRecommendations = new ArrayList<>(
                    this.graph.get(book).entrySet());
            sortedRecommendations.sort((a, b) -> 
                b.getValue().compareTo(a.getValue()));
            
            return sortedRecommendations.stream()
                     .limit(limit)
                     .collect(Collectors.toList());
        }
    }
}

File: data structures and algorithms/cs/05_002.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class MainClass
{
   public static void Main(string[] args)
   {
      // Main method code goes here
   }

   public class BookRecommendationGraph
   {
      private Dictionary<string, Dictionary<string, int>> graph;

      public BookRecommendationGraph()
      {
         this.graph = new Dictionary<string, Dictionary<string, int>>();
      }

      public void AddBookRelationship(string book1, string book2, int weight)
      {
         if (!this.graph.ContainsKey(book1))
         {
            this.graph[book1] = new Dictionary<string, int>();
         }
         if (!this.graph.ContainsKey(book2))
         {
            this.graph[book2] = new Dictionary<string, int>();
         }

         this.graph[book1][book2] = weight;
         this.graph[book2][book1] = weight;
      }

      public List<KeyValuePair<string, int>> GetRecommendations(string book, int limit)
      {
         if (!this.graph.ContainsKey(book))
         {
            return new List<KeyValuePair<string, int>>();
         }

         List<KeyValuePair<string, int>> sortedRecommendations = this.graph[book].ToList();
         sortedRecommendations.Sort((a, b) => b.Value.CompareTo(a.Value));

         return sortedRecommendations.Take(limit).ToList();
      }
   }
}
